keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -0,0 +1,847 @@
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+`PodSecurityContext.SupplementalGroups` is a list of groups applied to the first process run in each container, in addition to the container's primary GID, the `fsGroup` (if specified), and __group memberships defined in the container image for the UID__ of the container process. However, most cluster administrators and users would not recognize this behavior, especially regarding the groups defined in the container image.
+
+For example, if we have a Pod like this:
+```yaml
+spec:
+  # the primary UID will be uid=1000 in the container
+  securityContext: { runAsUser:1000, runAsGroup:1000, supplementalGroups:[50000]}
+  containers:
+  - image: some-image # uid=1000 belongs to gid=60000 in the image
+    sh: ["id"]
+```
+Then, groups of the first process of the container will be :
```

Comment:
I do not see this behavior in trivial testing - am I doing something wrong?

```
$ docker run -u 65534 ubuntu id
uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup)
```

This is correct - 65534 is the /etc/passwd group for UID 65534,

```
$ docker run -u 65534 --group-add 1 --group-add 2 ubuntu id
uid=65534(nobody) gid=65534(nogroup) groups=65534(nogroup),1(daemon),2(bin)
```

This is correct - I added groups 1 and 2, and the default group still exists.

```
$ docker run -u 65534:10 --group-add 1 --group-add 2 ubuntu id
uid=65534(nobody) gid=10(uucp) groups=10(uucp),1(daemon),2(bin)
```

I asked for a specific GID (runAsGroup equiv) and some extra.  The /etc/passwd group (65534) is not to be found.

So I seem to misunderstand the problem statement

Reply:
Thanks for the check. Yeah, I knew this docker's behavior. It's a kind of pitfall.

> I asked for a specific GID (runAsGroup equiv)

In [the official docker document of `USER` directive](https://docs.docker.com/engine/reference/builder/#user), 
(_note: `-u` option's semantics is the same with the directive_)

> Note that when specifying a group for the user, the user will have only the specified group membership. Any other configured group memberships will be ignored.

Although I'm not sure about the original intent of specifications of `runAsGroup` and `supplementalGroups`, this behavior is NOT the same as the current the APIs' implementations.

Reply:
Thanks for the check. Yeah, I know this docker's behavior. I think It's a kind of pitfall.

> I asked for a specific GID (runAsGroup equiv)

In [the official docker document of `USER` directive](https://docs.docker.com/engine/reference/builder/#user), 
(_note: `-u` option's semantics is the same with the directive_)

> Note that when specifying a group for the user, the user will have only the specified group membership. Any other configured group memberships will be ignored.

Although I'm not sure about the original intent of specifications of `runAsGroup` and `supplementalGroups`, this behavior is NOT the same as the current the APIs' implementations.

Reply:
Is it possible this is a difference between containerd/crio/docker ?

The OCI spec (https://github.com/opencontainers/image-spec/blob/main/config.md) says:

```
If group/gid is not specified, the default group and supplementary groups of the given user/uid in /etc/passwd from the container are applied.
```

To my reading, that says that adopting the from-passwd group is *wrong* when runAsGroup is specified.

IOW, this KEP feels like it is working around broken implementation(s) of OCI.

Am I missing something?

Reply:
> To my reading, that says that adopting the from-passwd group is wrong when runAsGroup is specified.

I think so.

> IOW, this KEP feels like it is working around broken implementation(s) of OCI.

Yes, I think this KEP can be said so.

However, I didn't confirm it in all the OCI implementations, but, specifying gid in [`USER` directive in Dockerfile](https://docs.docker.com/engine/reference/builder/#user:~:text=USER%F0%9F%94%97,will%20be%20ignored.) or in [`config.User` property in OCI image](https://github.com/opencontainers/image-spec/blob/main/config.md) would work as expected. 

I think the broken thing is `runAsGroup`/`supplementalGroups` API's implementations by the OCI implementations(at least containerd and crio).

Thus, this KEP proposes a new API to control the behavior because changing the previous behavior is a breaking-change and will make a big impact on users even if it was a bug.

Reply:
I'm not happy to add new APIs to work around spec-violating bugs.

I think we need to:

1) Confirm that this behavior exists in both containerd and crio.  (are there others we should check?)
2) File bugs against violators, citing OCI
3) Decide if they want to fix the bugs, fix the spec, or remain in violation
4) ONLY THEN can we decide if we need to add k8s level APIs

Reply:
> Confirm that this behavior exists in both containerd and crio. (are there others we should check?)

I already did on contained and crio. The reproduction code is [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups#readme) (this is also cited in the original issue https://github.com/kubernetes/kubernetes/issues/112879 which causes the KEP).

> File bugs against violators, citing OCI

I think this bug relates to CRI spec, not OCI image spec. Affected specs are 

- https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L321-L332
- https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1/api.proto#L869-L885

As I described [kubernetes/kubernetes#112879](https://github.com/kubernetes/kubernetes/issues/112879#:~:text=Root%20Cause%3A%20ambiguous%20definition), I think the root cause would be ambiguous API definition on both on Kubernetes API (`{Pod,Container}SecurityContext.{RunAsGroup, SupplementalGroups}`) and CRI spec (`Linux{Container,Sandbox}SecurityContext.{run_as_group, supplemental_groups}`).

I updated both API definition in https://github.com/kubernetes/kubernetes/pull/113047.

Reply:
I apologize for being a pain on this one - it just feels like an API we should not need, and I really want to understand why we do, if we do.  Also I have some history with supplemental groups, so I am shocked to see this.

I just tried to repro thru k8s - CRI to containerd

```
$ cat /tmp/p.yaml 
apiVersion: v1
kind: Pod
metadata:
  generateName: supplementalgroups-
spec:
  securityContext:
    runAsUser: 65534
    runAsGroup: 10
    supplementalGroups: [1,2]
  restartPolicy: Never
  containers:
  - image: ubuntu
    name: main
    command: ["bash", "-c", "id; true"]


$ k create -f /tmp/p.yaml
pod/supplementalgroups-5w9fr created

$ k logs supplementalgroups-5w9fr
uid=65534(nobody) gid=10(uucp) groups=10(uucp),1(daemon),2(bin)
```

That does not show the problem that you are supposedly fixing.

What am I doing wrong?  

Assuming I am doing something wrong:

https://github.com/kubernetes/kubernetes/issues/112879 immediately strikes out the "fix the bug" option as a breaking change.  I am usually on the side of not breaking things that work, but this is  a) a clear spec violation; b) a bug in one or more runtimes which would be a candidate to fix at those levels and then break us anyway; c) shocking behavior from a security POV.  I don't see the discussion about whether to actually fix this or not - did that discussion happen?

Before we merge new API surface to work around this, I'd like to see code citations - where EXACTLY is this happening?  Some piece of code must be looking at the OCI image, extracting the group ID, and adding it to supplementalGroups - I'd like to hear from the owner(s) of that code about why they think this is correct interpretation of OCI.

Finally, if we do fix it, I'd also want to see something at those code-sites that prevents future people from fixing the bug - because it seems pretty clearly a bug to me, and broken things have a tendency to get fixed.

But first and foremost - I can't reproduce this, and it worries me.

Reply:
> I apologize for being a pain on this one - it just feels like an API we should not need, and I really want to understand why we do, if we do. Also I have some history with supplemental groups, so I am shocked to see this.

Thank you for this kind comment. I also didn't want to change Kubernetes API but to just fix the original issue. I think we're on the same page. I'm really happy to discuss what the best way is. Yeah, I was so shocked when I found the behavior, too.

> But first and foremost - I can't reproduce this, and it worries me.

Hmmm, It is strange (but I see why this happens now, see below). I also confirmed [the reproduction code is still effective](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups#readme). It needs an explicit member list part in the /etc/group entries.

```shell
# using a custom image:
#   user alice belongs to alice(self group) and bypassed-group
#   Dockerfile: https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/blob/reproduce-bypass-supplementalgroups/Dockerfile
$ docker run everpeace/kubernetes-issue-112879 cat /etc/group
...
bypassed-group:x:50000:alice
alice:x:1000:

$ cat /tmp/p.yaml 
apiVersion: v1
kind: Pod
metadata:
  generateName: bypass-supplementalgroups-pod-
spec:
  securityContext:
    runAsUser: 1000 
    runAsGroup: 1000
    supplementalGroups: [60000]
  restartPolicy: Never
  containers:
  - image: everpeace/kubernetes-issue-112879
    name: main
    command: ["bash", "-c", "id; true"]
    imagePullPolicy: Never

$ k create -f /tmp/p.yaml
pod/bypass-supplementalgroups-pod-nvx7l created

# 50000(defined in /etc/groups) appears!!
$ k logs pod/bypass-supplementalgroups-pod-nvx7l
uid=1000(alice) gid=1000(alice) groups=1000(alice),50000(bypassed-group),60000
```

> Before we merge new API surface to work around this, I'd like to see code citations - where EXACTLY is this happening?

In contained, I think the below code causes the situation

- `containerSpecOpts` is called in `CreateContainer`: https://github.com/containerd/containerd/blob/v1.6.16/pkg/cri/server/container_create.go#L230
- in `containerSpecOpts`, configuring OCI container spec, especially [user configuration](https://github.com/opencontainers/runtime-spec/blob/main/config.md#user): https://github.com/containerd/containerd/blob/v1.6.16/pkg/cri/server/container_create_linux.go#L325-L351
- in coantainerSpecOpts, even if `runAsGroup` is specified, scanning /etc/passwd by calling `customopts.WithAdditionalGIDs`: https://github.com/containerd/containerd/blob/v1.6.16/pkg/cri/server/container_create_linux.go#L351
- `customopts.WithAdditionalGIDs` just delegates to `oci.WithAdditionalGIDs`: https://github.com/containerd/containerd/blob/v1.6.16/pkg/cri/opts/spec_linux.go#L54
- `oci.WithAdditionalGIDs` actually scan's /etc/passwd and resolves additional gids: https://github.com/containerd/containerd/blob/v1.6.16/oci/spec_opts.go#L738:6
- _They explicitly exclude the group having the same name with the username. This is why nobody(65534) doesn't appear in your case_: https://github.com/containerd/containerd/blob/v1.6.16/oci/spec_opts.go#L763-L766

> https://github.com/kubernetes/kubernetes/issues/112879 immediately strikes out the "fix the bug" option as a breaking change. I am usually on the side of not breaking things that work, but this is a) a clear spec violation; b) a bug in one or more runtimes which would be a candidate to fix at those levels and then break us anyway; c) shocking behavior from a security POV. I don't see the discussion about whether to actually fix this or not - did that discussion happen?

Yes, as described ["Is this a security issue??" section in kubernetes#112879](https://github.com/kubernetes/kubernetes/issues/112879#:~:text=Is%20this%20a%20security%20issue%3F%3F) we firstly reported this behavior to hackerone.com(`#1688374`) by following [Kubernetes Security and Disclosure Information](https://kubernetes.io/docs/reference/issues-security/security/). Kubernetes Security Response Committee responded that this behavior "works as intended," and they recommended discussing how to handle this behavior in a public k/k issue.

And then, I posted the issue in the sig-node meetings, and sig [recommended me](https://github.com/kubernetes/kubernetes/issues/112879#:~:text=How%20to%20resolve%20this%20issue%3F) to improve API definition first and discuss how to fix the issue(it's current status).

---

> https://github.com/kubernetes/kubernetes/issues/112879 immediately strikes out the "fix the bug" option as a breaking change

Then, once you confirmed this issue is "the bug", the next step would share the bug and spec of `runAsGroups` with major CRI runtime communities: contained, cri-o, and others?

Reply:
OK, let's try to repro it again:

```
$ docker run -ti ubuntu
root@624168bd5677:/# echo thockin:x:9376:9376:root:/tmp:/bin/bash >>/etc/passwd
root@624168bd5677:/# echo thockin:x:9376: >> /etc/group
root@624168bd5677:/# echo extra:x:9377:thockin >> /etc/group
root@624168bd5677:/# echo extra2:x:9378:thockin >> /etc/group
root@624168bd5677:/# exit

$ docker commit 624168bd5677
sha256:0c6fbb0fbd1bc1a2d1e0011c78ccc43869e997f7f633fa3dc9acce52aa2c352f

$ docker tag sha256:0c6fbb0fbd1bc1a2d1e0011c78ccc43869e997f7f633fa3dc9acce52aa2c352f thockin/supplemental

$ docker run -ti -u thockin thockin/supplemental id
uid=9376(thockin) gid=9376(thockin) groups=9376(thockin),9377(extra),9378(extra2)

$ docker run -ti -u thockin:thockin thockin/supplemental id
uid=9376(thockin) gid=9376(thockin) groups=9376(thockin)

$ docker run -ti -u thockin:10 thockin/supplemental id
uid=9376(thockin) gid=10(uucp) groups=10(uucp)
```

docker itself seems OK

Let's try k8s:

```
$ cat /tmp/p.yaml 
apiVersion: v1
kind: Pod
metadata:
  generateName: supplementalgroups-
spec:
  securityContext:
    runAsUser: 9376
    runAsGroup: 10
  restartPolicy: Never
  containers:
  - image: thockin/supplemental
    name: main
    command: ["bash", "-c", "id; true"]

$ k create -f /tmp/p.yaml -o name
pod/supplementalgroups-29n4q

$ k logs pod/supplementalgroups-29n4q
uid=9376(thockin) gid=10(uucp) groups=10(uucp),9377(extra),9378(extra2)
```

Aha!  That is different.

Let's try nerdctl:

```
$ nerdctl run --net=none -ti -u thockin thockin/supplemental id
uid=9376(thockin) gid=9376(thockin) groups=9376(thockin),9377(extra),9378(extra2)

$ /var/run/docker/nerdctl run --net=none -ti -u thockin:thockin thockin/supplemental id
uid=9376(thockin) gid=9376(thockin) groups=9376(thockin)

$ /var/run/docker/nerdctl run --net=none -ti -u thockin:10 thockin/supplemental id
uid=9376(thockin) gid=10(uucp) groups=10(uucp)
```

That seems correct.  So the issue is in containerd CRI impl - containerd gets it right on one path and wrong on another.  I don't want to know what pathology leads to that...

Thanks for your patience.  I think I am now caught up to what you wrote.

I think it is a security issue, but not a major one, probably.  More importantly, I think we have a divergence from the OCI spec, and that's the important thing.  I do not think it is an intentional or meaningful divergence (at the k8s layer).  The questions I have then are:

* Do containerd and CRI-O owners think the current behavior is OCI compliant?
* Do containerd and CRI-O owners think the current behavior is a feature (because they interpret k8s as demanding it) or a a bug (oops)?
* If they agree it's not OCI and probably not what was intended, should we codify it into k8s API (compatibility) or fix it?
* If they agree not to fix it, do they have sufficient tests and docs/comments in place that nobody will "accidentially" fix it later?

Reply:
@estesp if you have a moment wrt containerd

@jonjohnsonjr if you have a moment - do you have an opinion about whether this is OCI compliant?  I think the k8s/CRI behavior is wrong, but the OCI spec doesn't QUITE say so.

Reply:
I'd defer to @vbatts or @tianon -- image-spec -> runtime-spec is not in my wheelhouse.

Reply:
@thockin Thanks for digging into this. I think that we should fix the CRI runtimes. I assumed docker behaved this way and the CRIs were retaining that behavior but that isn't the case.  I am :+1: to fixing CRI runtimes for this.

Reply:
cc: @mikebrow for containerd.

Reply:
@thockin's understanding (and thorough testing, as usual :heart:) matches my own :+1:

However, FWIW from the OCI perspective, this _is_ defined in the image-spec (the runtime-spec only defines the object that includes the final `uid`/`gid`/`additionalGids`; https://github.com/opencontainers/runtime-spec/blob/5cfc4c32c234bf00ef9e4bdee0576af25dc23629/config.md#posix-platform-user vs `config.User` in https://github.com/opencontainers/image-spec/blob/2fb996805b3734779bf9a3a84dc9a9691ad7efdd/config.md).

Reply:
The image spec says "If group/gid is not specified, the default group and supplementary groups of the given user/uid in /etc/passwd from the container are applied."

It doesn't say anything about /etc/group (that's a nitpick) and it doesn't say that supplementary groups from the container will be (ignored | merged) when group or gid is specified.

I interpretted it to mean "ignored" when GID is specified, and Docker and containerd defailt impl agrees with me.  It seems that k8s and CRI were interpreting it as "merged".  Maybe the spec should say explicitly what is correct.

Reply:
Yeah, I think the original intent there was that "supplementary groups" was "the groups from /etc/group" since passwd and group are usually a pair, but that'd be a fair update to clarify in the spec IMO.

Reply:
The image spec statement covers  the interpretation of config.User in the image metadata (derived from USER in a Dockerfile). When we are explicitly overriding that by setting RunAsUser, RunAsGroup or SupplementalGroups in SecurityContext that is a runtime override and I think that behavior should be defined in k8s/CRI spec and backed by additional tests.

Reply:
I'm not saying we shouldn't define it, just that it would be surprising to me if we defined it _differently_ than OCI.

That said, it's a breaking change for us (in theory - I doubt very much it will impact anyone), so maybe we have to.

Reply:
Thank you guys for the discussion! No matter whether the implementation gap is intended or not intended, I also agree the gap should be fixed in some way. So, wrapping up, I feel we agreed on the below:

- CRI runtimes(at least containerd, crio and more) should be fixed, and
- We should introduce clear k8s/CRI spec about this (i.e. keep reviewing the KEP for how to fix k8s/CRI).

Is my understanding correct?

---

> I think that behavior should be defined in k8s/CRI spec and backed by additional tests.

Regarding CRI tests, [kubernetes-sigs/cri-tools](https://github.com/kubernetes-sigs/cri-tools) ships [`critest` (CRI validation testing tool)](https://github.com/kubernetes-sigs/cri-tools/blob/master/docs/validation.md) for CRI implementors. I think adding the test cases to the tool would be a standard way for it.

By the way, I recently added SupplementalGroups behavioral tests which assert the _current_ behavior (incompatible with OCI image spec) in https://github.com/kubernetes-sigs/cri-tools/pull/1005. I will also add test cases for `SupplementalGroupsPolicy` if the KEP is approved.